compaction/compaction_manager.cc

```diff
@@ -676,12 +676,16 @@ sstables_task_executor::~sstables_task_executor() {
 }
 
 future<compaction_manager::compaction_stats_opt> compaction_task_executor::run_compaction() noexcept {
-    try {
-        _compaction_done = do_run();
-        return compaction_done();
-    } catch (...) {
-        return current_exception_as_future<compaction_manager::compaction_stats_opt>();
-    }
+    return do_run().then_wrapped([&] (auto f) {
+        try {
+            auto res = f.get();
+            _compaction_done.set_value(res);
+            return make_ready_future<compaction_manager::compaction_stats_opt>(res);
+        } catch (...) {
+            _compaction_done.set_exception(std::current_exception());
+            return current_exception_as_future<compaction_manager::compaction_stats_opt>();
+        }
```

Comment:
I'm not a big fan of using f.get() and handling the exception when it holds one.
We could check for f.failed() as follows, but I'm not sure how much simpler it is.
I guess it's ok to code it the way you did, especially since this ins't a performance sensitive path.

```c++
    if (f.failed()) {
        auto ex = f.get_exception();
        _compaction_done.set_exception(ex);
        return make_exception_future<compaction_manager::compaction_stats_opt>(std::move(ex));
    }
    auto res = f.get();
    _compaction_done.set_value(res);
    return make_ready_future<compaction_manager::compaction_stats_opt>(std::move(res));
```